### PR TITLE
Add methods to move ByteStrings underlying bytes

### LINKED ISSUE #9655: Add a method to transfer ownership of the underlying Vec<u8> for ByteString
And perhaps create a new Vec<u8> in its place after ownership was taken.

Code: `components/script/dom/bindings/str.rs`

### PR DESCRIPTION
**Pull request by dlrobertson**
Add methods to move the underlying `Vec<u8>` for `ByteString`.

I saw this as at least two methods. One to "move and replace with and empty Vec<u8> (`bytes`), and one to take ownership of the whole object (`own_bytes`). I typically also don't like adding methods with out unit tests. If you think they're unnecessary, just let me know.

As always, please let me know if you have any comments, critiques, or nits!

Fixes #9655

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9684)

### CODE REVIEW

**Comment by danlrobertson** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/script/dom/bindings/str.rs, line 46 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9684#-KAlERu0O5BUc3SuBAGP:-KAlERu0O5BUc3SuBAGQ:558965142) ([raw file](https://github.com/servo/servo/blob/7f332515fd00849491ed74100d4b57efea93324d/components/script/dom/bindings/str.rs#L46)):*
Saw the function was using the older method, so I simplified it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9684)*

**Comment by KiChjang** [comment]
Wow, this is pretty good work, thanks for your contribution!

@bors-servo r+

File: components/script/dom/bindings/str.rs [hunk 1]
```diff
@@ -28,10 +29,21 @@ impl ByteString {
         str::from_utf8(&vec).ok()
     }
 
+    /// Returns ownership of the underlying Vec<u8> the ByteString is
+    /// unusable afterwards as it has been moved
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: `Takes ownership of the underlying Vec<u8>. The ByteString is unusable afterwards.`

File: components/script/dom/bindings/str.rs [hunk 2]
```diff
@@ -28,10 +29,21 @@ impl ByteString {
...
+    pub fn own_bytes(self) -> Vec<u8> {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: I'm thinking that this should be instead named `into_vec` to be consistent with `std` namings.

File: components/script/dom/bindings/str.rs (hunk 1 above)
Comment (danlrobertson) [inline components/script/dom/bindings/str.rs]: Much better documentation!

File: components/script/dom/bindings/str.rs (hunk 2 above)
Comment (danlrobertson) [inline components/script/dom/bindings/str.rs]: Ah! Good point! I thought about something like `drain_bytes`, but we'll go with `into_bytes`.

File: components/script/dom/bindings/str.rs [hunk 3]
```diff
@@ -24,14 +25,24 @@ impl ByteString {
...
     /// otherwise.
     pub fn as_str(&self) -> Option<&str> {
-        let ByteString(ref vec) = *self;
-        str::from_utf8(&vec).ok()
+        str::from_utf8(&self.0).ok()
+    }
+
+    /// Takes ownership of the underlying Vec<u8>. The ByteString is
+    /// unusable afterwards.
+    pub fn own_bytes(self) -> Vec<u8> {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: I meant this method to be renamed `into_bytes`.

File: components/script/dom/bindings/str.rs [hunk 4]
```diff
@@ -24,14 +25,24 @@ impl ByteString {
...
+        self.0
+    }
+
+    /// Returns ownership of the underlying Vec<u8> and copies an empty
+    /// vec in its place
+    pub fn into_bytes(&mut self) -> Vec<u8> {
```
Comment (KiChjang) [inline components/script/dom/bindings/str.rs]: `bytes` is fine.

Comment (danlrobertson) [inline components/script/dom/bindings/str.rs]: Oh, my bad! One sec

**Comment by KiChjang** [comment]
@bors-servo retry
* #9033

**Comment by KiChjang** [comment]
@bors-servo retry
* #9092

**Pull request merged**